Reply indicator

l1n opened the pull request:
Closes #1154. This will have performance impact, probably better to merge after postgres is moved. To mitigate somewhat, the indicator is defaulted to off.

Marri commented on app/helpers/application_helper.rb:
```diff
@@ -252,4 +252,20 @@ def allowed_boards(obj, user)
+      @hiatused_exist = true if hiatused.count > 0
+    end
+
+    return @posts.count
```
None of these should be instance variables; among other reasons, we have a number of controllers which set a different @posts variable and they'll conflict. You're not using them anywhere else, so they can safely drop the @.

Marri commented on app/views/layouts/application.haml:
```diff
@@ -111,6 +111,10 @@
               %li= link_to 'Unread', unread_posts_path
               %li= link_to 'Replies Owed', owed_posts_path
+              - if current_user.replies_owed_indicator
+                - $owed = get_owed
```
Don't need the $ here; that makes it a global, which we don't especially need it to be doing.

Marri commented on app/helpers/application_helper.rb:
```diff
@@ -252,4 +252,20 @@ def allowed_boards(obj, user)
+  def get_owed
+    ids = Post::Author.where(user_id: current_user.id, can_owe: true).group(:post_id).pluck(:post_id)
+    @posts = Post.where(id: ids)
+    drafts = ReplyDraft.where(post_id: @posts.select(:id)).where(user: current_user).pluck(:post_id)
```
You should be able to use `ids` here instead of `@posts.select(:id)` to make it more efficient

Marri commented on app/helpers/application_helper.rb:
```diff
@@ -252,4 +252,20 @@ def allowed_boards(obj, user)
+    solo = Post::Author.where(post_id: ids).group(:post_id).having('count(post_id) < 2').pluck(:post_id)
+    @posts = @posts.where.not(last_user: current_user).or(@posts.where(id: (drafts + solo).uniq))
+    @posts = @posts.where.not(status: [:complete, :abandoned])
+    hiatused = @posts.hiatus.or(@posts.where('tagged_at < ?', 1.month.ago))
```
You don't need this (I assume it's just a copy/paste artifact)

Marri commented on app/helpers/application_helper.rb:
```diff
@@ -252,4 +252,20 @@ def allowed_boards(obj, user)
+
+    if current_user.hide_hiatused_tags_owed?
+      @posts = @posts.where.not(status: :hiatus).where('tagged_at > ?', 1.month.ago)
+      @hiatused_exist = true if hiatused.count > 0
```
Ditto copy/paste artifact (the rest of the if block is fine)

Marri commented on app/views/layouts/application.haml:
```diff
@@ -111,6 +111,10 @@
               %li= link_to 'Replies Owed', owed_posts_path
+              - if current_user.replies_owed_indicator
+                - $owed = get_owed
+                - if $owed
```
You should probably check > 0 here; despite also being C-based, Ruby (unlike Python) generally treats 0 as truthy.

Marri commented on app/views/users/edit.haml:
```diff
@@ -80,6 +80,9 @@
           = f.label :hide_hiatused_tags_owed, "Hide hiatused threads"
+          %br
+          = f.check_box :replies_owed_indicator, class: 'width-15 vmid'
+          = f.label :replies_owed_indicator, "Show replies owed count"
```
Might want an extra clarifying "in top menu bar" or something here

Marri reviewed (changes requested):
Suggested various tweaks mostly to fix copy paste artifacts; also as a general comment needs tests

Commit:
Reply-owed count indicator in the top nav

Adds an opt-in badge that shows the number of posts the user owes
a reply to next to the "Replies Owed" nav link. Off by default;
toggled from the user settings page.

Addresses the review feedback on #1503:
- Helper takes the user as a local parameter instead of leaning
  on instance variables in a helper method.
- No globals — the count is a local in the haml.
- `ReplyDraft.where(post_id: ids, user: ...)` uses the existing
  `ids` array instead of an inner `posts.select(:id)` subquery.
- Drops the dead `hiatused = ...` / `@hiatused_exist = ...`
  lines that were copy-paste artifacts from `PostsController#owed`
  (they were never read in the helper context).
- Explicit `> 0` check on the count in haml; relies less on
  Ruby's truthiness rules for an integer.
- Settings label clarifies "in top menu bar" so the toggle's
  effect is obvious.
- Adds `spec/helpers/application_helper_spec.rb#replies_owed_count`
  covering nil user, no owed, has owed, complete/abandoned, and
  hiatused-hidden cases.

Fixes #1154.
